### PR TITLE
Add intersection,intersectionWith for Data.Map

### PR DESCRIPTION
**Pull request by karshan**
Not sure why these are missing.

### CODE REVIEW

File: src/Data/Map/Internal.purs [hunk 1]
```diff
@@ -628,6 +630,24 @@ union = unionWith const
...
+  go as@(Cons (Tuple k1 a) ass) bs@(Cons (Tuple k2 b) bss) m =
+    case compare k1 k2 of
+         LT -> go ass bs m
+         EQ -> go ass bss (insert k1 (f a b) m)
+         GT -> go as bss m
+
+-- | Compute the intersection of two maps, preferring values from the first map in the case
+-- | of duplicate keys.
+intersection :: forall k a b. Ord k => Map k a -> Map k b -> Map k a
+intersection = intersectionWith (\a b -> a)
```
Comment (hdgarrood) [inline src/Data/Map/Internal.purs]: Minor nit but I think it would be better to use `const` here.

File: test/Test/Data/Map.purs [hunk 2]
```diff
@@ -223,6 +223,23 @@ mapTests = do
...
+  quickCheck $ \(TestMap m1) (TestMap m2) k ->
+    M.lookup (smallKey k) (M.intersection (m1 :: M.Map SmallKey Int) (m2 :: M.Map SmallKey Int)) == (case M.lookup k m2 of
+      Nothing -> Nothing
+      Just v -> M.lookup k m1) <?> ("m1: " <> show m1 <> ", m2: " <> show m2 <> ", k: " <> show k <> ", v1: " <> show (M.lookup k m1) <> ", v2: " <> show (M.lookup k m2) <> ", intersection: " <> show (M.intersection m1 m2))
+
+  log "Intersection is idempotent"
+  quickCheck $ \(TestMap m1) (TestMap m2) -> ((m1 :: M.Map SmallKey Int) `M.intersection` m2) == ((m1 `M.intersection` m2) `M.intersection` (m2 :: M.Map SmallKey Int))
+
+  log "intersectionWith"
+  for_ [Tuple (+) 0, Tuple (*) 1] $ \(Tuple op ident) ->
```
Comment (hdgarrood) [inline test/Test/Data/Map.purs]: Is `ident` being used here?

**Review by hdgarrood — COMMENTED** [review]
Looking good so far, just a couple of minor notes. I think adding these makes sense; `intersection` seems a little weird to me in terms of the amount of information it discards, and I don’t think I’ve ever wanted it, but `intersectionWith` seems more sensible at least, and also both of these functions already exist in Haskell’s `containers` library anyway.

File: test/Test/Data/Map.purs (hunk 2 above)
Comment (karshan) [reply test/Test/Data/Map.purs]: Nope, artifact from unionWith test this is based of. Removed.

File: src/Data/Map/Internal.purs (hunk 1 above)
Comment (karshan) [reply src/Data/Map/Internal.purs]: Ah yes ofcourse, fixed.

**Review by hdgarrood — APPROVED** [review]
Nice, thank you!

**Pull request merged**